synapseadmin/roomapi.go

```diff
@@ -0,0 +1,167 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package synapseadmin
+
+import (
+	"context"
+	"net/http"
+	"maunium.net/go/mautrix"
+	"maunium.net/go/mautrix/id"
+	"strconv"
+)
+
+type ReqListRoom struct {
+	Alias	string
+	From	int
+	Limit   int
+}
+
+func (cli *Client) NewReqListRoom() ReqListRoom {
+   req := ReqListRoom{}
+   req.Alias = ""
+   req.From = 0
+   req.Limit = 100
+   return req
+}
+
+func (req *ReqListRoom) BuildQuery() map[string]string {
+	query := map[string]string{
+		"limit": strconv.Itoa(req.Limit),
+	}
+	if req.Alias != "" {
+		query["search_term"] = req.Alias
+	}
+	query["from"] = strconv.Itoa(req.From)
+	return query
+}
+
+//  Get room info based on alias
+//
+// https://matrix-org.github.io/synapse/latest/admin_api/rooms.html
+func (cli *Client) ListRoom(ctx context.Context, req ReqListRoom) (*mautrix.RoomsResponse ,error) {
+	var resp mautrix.RoomsResponse
+	var reqURL string
+	reqURL = cli.BuildURLWithQuery(mautrix.SynapseAdminURLPath{"v1", "rooms"}, req.BuildQuery())//map[string]string{"search_term": alias})
+	_, err := cli.MakeFullRequest(ctx, mautrix.FullRequest{
+		Method:       http.MethodGet,
+		URL:          reqURL,
+		ResponseJSON: &resp,
+	})
+	return &resp, err
+}
+
+// ReqDeleteRoom is the request content for Client.DeleteRoom.
+type ReqDeleteRoom struct {
+	// Room Id to delete
+	RoomId string `json:"room_id"`
+	// Default to true: remove all traces from database
+	Purge bool `json:"purge"`
+}
+
+//  Delete Room based on Id
+//
+// https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#version-2-new-version
+func (cli *Client) DeleteRoom(ctx context.Context, req ReqDeleteRoom) error {
+	reqURL := cli.BuildAdminURL("v2", "rooms", req.RoomId)
+	_, err := cli.MakeFullRequest(ctx, mautrix.FullRequest{
+		Method:       http.MethodDelete,
+		URL:          reqURL,
+		RequestJSON: &req,
+	})
+	return err
+}
+
+// Get Room Members based on Room Id
+//
+// https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#version-2-new-version
+func (cli *Client) RoomMembers(ctx context.Context, RoomId string) (*mautrix.RoomsMembersResponse, error) {
+	reqURL := cli.BuildAdminURL("v1", "rooms", RoomId, "members")
+	var resp mautrix.RoomsMembersResponse
+	_, err := cli.MakeFullRequest(ctx, mautrix.FullRequest{
+		Method:       http.MethodGet,
+		URL:          reqURL,
+		ResponseJSON: &resp,
+	})
+	return &resp, err
+}
+
+// ReqRoomAdmin is the request content for Client.RoomAdmin.
+type ReqRoomAdmin struct {
+	// User Id to make admin
+	UserId id.UserID `json:"user_id"`
+	// Room Id or alias
+	Room string `json:"room"`
+}
+
+//  Make admin a Room's user based on Id
+//
+// https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#make-room-admin-api
+func (cli *Client) RoomAdmin(ctx context.Context, req ReqRoomAdmin) error {
+	reqURL := cli.BuildAdminURL("v1", "rooms", req.Room, "make_room_admin")
```

Comment:
In general, path parameters should go in the function arguments and request body should go in a struct. Don't mix them unless there's a good reason to.

(this applies to a few other methods too)